framework/src/userobject/NearestPointAverage.C

```diff
@@ -0,0 +1,98 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "NearestPointAverage.h"
+
+registerMooseObject("MooseApp", NearestPointAverage);
+
+template <>
+InputParameters
+validParams<NearestPointAverage>()
+{
+  InputParameters params =
+      nearestPointBaseValidParams<ElementAverageValue, ElementVariableVectorPostprocessor>();
+
+  params.addClassDescription(
+      "Compute element variable averages for nearest-point based subdomains");
+
+  return params;
+}
+
+NearestPointAverage::NearestPointAverage(const InputParameters & parameters)
+  : NearestPointBase<ElementAverageValue, ElementVariableVectorPostprocessor>(parameters),
+    _np_post_processor_values(declareVector("np_post_processor_values"))
+{
+  _np_post_processor_values.resize(_user_objects.size());
+}
+
+Real
+NearestPointAverage::spatialValue(const Point & point) const
+{
+  unsigned int i = nearestPointIndex(point);
+
+  if (i >= _np_post_processor_values.size())
+    mooseError("The vector length of vector post processor is ",
+               _np_post_processor_values.size(),
+               " but nearestPointIndex() return ",
+               i);
+
+  return _np_post_processor_values[i];
+}
+
+Real
+NearestPointAverage::userObjectValue(unsigned int i) const
+{
+  if (i >= _np_post_processor_values.size())
+    mooseError("The vector length of vector post processor is ",
+               _np_post_processor_values.size(),
+               " but you pass in ",
+               i);
+
+  return _np_post_processor_values[i];
+}
+
+void
+NearestPointAverage::finalize()
+{
+  if (_user_objects.size() != _np_post_processor_values.size())
+    mooseError("The vector length of the vector postprocessor ",
+               _np_post_processor_values.size(),
+               " is different from the number of user objects ",
+               _user_objects.size());
+
+  unsigned int i = 0;
+  for (auto & user_object : _user_objects)
+  {
+    user_object->finalize();
+    _np_post_processor_values[i++] = user_object->getValue();
+  }
+}
+
+unsigned int
+NearestPointAverage::nearestPointIndex(const Point & p) const
+{
+  unsigned int closest = 0;
+  Real closest_distance = std::numeric_limits<Real>::max();
+
+  for (auto it : Moose::enumerate(_points))
```

Comment:
can the closest _point be off-domain? What happens for distributed mesh? 

Had a quick look, did not find which object handles that part. Please just point me in the right direction if you've looked

Reply:
The `_points` are provided by the user as a text file - they don't have any relation to how the mesh is parallelized. Locally, I created a test with a single `point` (so all nodes in the domain should be closest to the same x, y, z and contribute to the same "score") that works with a distributed mesh. I believe this demonstrates correct distributed mesh behavior because a partial average is summed across distributed parts of the mesh.

In Cardinal, we also have several tests of our OpenMC wrapping that have been tested for both distributed and replicated meshes (which include the `NearestPointAverage` in those inputs).

Are there other situations I should test as well?

Reply:
These tests are good. 
I ll go digging for the logic in the code. sorry for the delay

Reply:
Yeah I think it happens somewhere in the base classes. And haha no need to apologize, I only opened this yesterday! This is a fast review :)